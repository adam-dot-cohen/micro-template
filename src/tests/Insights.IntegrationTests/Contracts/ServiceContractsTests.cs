using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using Google.Protobuf.Reflection;
using Grpc.Core;
using Laso.Insights.IntegrationTests.Extensions;
using Laso.IntegrationEvents;
using Xunit;

namespace Laso.Insights.IntegrationTests.Contracts
{
    public class ServiceContractsTests
    {
        private static readonly ICollection<Type> AllReferencedTypes = Directory.GetFiles(AppDomain.CurrentDomain.BaseDirectory, "*.dll")
            .Select(AssemblyName.GetAssemblyName)
            .Where(x => x.Name?.StartsWith("Laso") == true)
            .Select(Assembly.Load)
            .SelectMany(x => x.GetTypes())
            .ToList();

        [Fact]
        public void Proto_files_should_not_have_breaking_changes()
        {
            static string GetType(FieldDescriptor field)
            {
                return field.FieldType switch
                {
                    FieldType.Message => field.MessageType.FullName,
                    FieldType.Enum => field.EnumType.FullName,
                    FieldType.Group => throw new NotSupportedException(),
                    _ => field.FieldType.ToString()
                };
            }

            AllReferencedTypes
                .Select(x => x.GetCustomAttribute<BindServiceMethodAttribute>(false)?.BindType)
                .Where(x => x != null)
                .Select(x => ((ServiceDescriptor) x.GetProperty("Descriptor").GetValue(null)).File)
                .Distinct(x => x.Name)
                .ToList()
                .To(x => new
                {
                    Methods = x
                        .SelectMany(y => y.Services, (y, z) => new { Proto = y.Name, Service = z })
                        .SelectMany(y => y.Service.Methods, (y, z) => new
                        {
                            y.Proto,
                            Service = y.Service.FullName,
                            z.Name,
                            InputType = z.InputType.FullName,
                            OutputType = z.OutputType.FullName
                        }),
                    Fields = x
                        .SelectMany(y => y.MessageTypes, (y, z) => new { Proto = y.Name, MessageType = z })
                        .SelectMany(y => y.MessageType.Fields.InFieldNumberOrder(), (y, z) => new
                        {
                            y.Proto,
                            MessageType = y.MessageType.FullName,
                            z.Name,
                            Type = GetType(z),
                            z.FieldNumber,
                            z.IsRequired,
                            z.IsRepeated,
                            //z.IsPacked, This seems to be bugged
                            z.IsMap,
                            ContainingOneof = z.ContainingOneof?.Name
                        }),
                    EnumValues = x
                        .SelectMany(y => y.EnumTypes, (y, z) => new { Proto = y.Name, EnumType = z })
                        .SelectMany(y => y.EnumType.Values, (y, z) => new
                        {
                            y.Proto,
                            EnumType = y.EnumType.FullName,
                            z.Name,
                            z.Number
                        })
                })
                .ApproveIf(x => x.Approved.Methods.All(y => x.Received.Methods.Contains(y))
                                && x.Approved.Fields.All(y => x.Received.Fields.Contains(y))
                                && x.Approved.EnumValues.All(y => x.Received.EnumValues.Contains(y))
                                && x.Received.EnumValues.All(y => x.Approved.EnumValues.Contains(y)));
        }

        [Fact]
        public void Integration_events_should_not_have_breaking_changes()
        {
            static void VisitMessageTypes(Type type, ICollection<Type> messageTypes, ICollection<Type> enumTypes)
            {
                if (messageTypes.Contains(type))
                    return;

                messageTypes.Add(type);

                type.GetProperties()
                    .Select(x => x.PropertyType.GetNonNullableType())
                    .ForEach(x =>
                    {
                        //Add more "primitive" types here
                        if (x.IsPrimitive || x == typeof(string) || x == typeof(DateTime) || x == typeof(Guid) || x == typeof(DateTimeOffset))
                            return;

                        if (x.IsEnum)
                            enumTypes.Add(x);

                        VisitMessageTypes(x.Closes(typeof(IEnumerable<>), out var args) ? args.First()[0] : x, messageTypes, enumTypes);
                    });
            }

            AllReferencedTypes
                .Where(x => typeof(IIntegrationEvent).IsAssignableFrom(x) && !x.IsInterface && !x.IsAbstract)
                .ToList()
                .To(x =>
                {
                    var messageTypes = new HashSet<Type>();
                    var enumTypes = new HashSet<Type>();

                    x.ForEach(y => VisitMessageTypes(y, messageTypes, enumTypes));

                    return new
                    {
                        Fields = messageTypes.SelectMany(y => y.GetProperties(), (y, z) => new
                        {
                            MessageType = y.FullName,
                            z.Name,
                            Type = GetTypeName(z.PropertyType)
                        }),
                        EnumValues = enumTypes.SelectMany(y => Enum.GetValues(y).Cast<Enum>(), (y, z) => new
                        {
                            EnumType = y.FullName,
                            Name = z.ToString(),
                            Value = z.GetValue()
                        })
                    };
                })
                .ApproveIf(x => x.Approved.Fields.All(y => x.Received.Fields.Contains(y))
                                && x.Approved.EnumValues.All(y => x.Received.EnumValues.Contains(y))
                                && x.Received.EnumValues.All(y => x.Approved.EnumValues.Contains(y)));
        }

        private static string GetTypeName(Type type)
        {
            if (type != typeof(string) && type.Closes(typeof(IEnumerable<>), out var args))
               return $"List<{args.First()[0].FullName}>";

            return type.FullName;
        }
    }
}
