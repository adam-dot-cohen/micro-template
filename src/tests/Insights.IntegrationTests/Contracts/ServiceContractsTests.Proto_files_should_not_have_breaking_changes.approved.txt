{
  "Methods": [
    {
      "Proto": "Protos/catalog.api.v1.proto",
      "Service": "catalog.api.v1.Greeter",
      "Name": "SayHello",
      "InputType": "catalog.api.v1.HelloRequest",
      "OutputType": "catalog.api.v1.HelloReply"
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "Service": "identity.api.v1.Partners",
      "Name": "CreatePartner",
      "InputType": "identity.api.v1.CreatePartnerRequest",
      "OutputType": "identity.api.v1.CreatePartnerReply"
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "Service": "identity.api.v1.Partners",
      "Name": "DeletePartner",
      "InputType": "identity.api.v1.DeletePartnerRequest",
      "OutputType": "identity.api.v1.DeletePartnerReply"
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "Service": "identity.api.v1.Partners",
      "Name": "GetPartners",
      "InputType": "identity.api.v1.GetPartnersRequest",
      "OutputType": "identity.api.v1.GetPartnersReply"
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "Service": "identity.api.v1.Partners",
      "Name": "GetPartner",
      "InputType": "identity.api.v1.GetPartnerRequest",
      "OutputType": "identity.api.v1.GetPartnerReply"
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "Service": "provisioning.api.v1.Partners",
      "Name": "ProvisionPartner",
      "InputType": "provisioning.api.v1.ProvisionPartnerRequest",
      "OutputType": "provisioning.api.v1.ProvisionPartnerReply"
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "Service": "provisioning.api.v1.Partners",
      "Name": "RemovePartner",
      "InputType": "provisioning.api.v1.RemovePartnerRequest",
      "OutputType": "provisioning.api.v1.RemovePartnerReply"
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "Service": "provisioning.api.v1.Partners",
      "Name": "GetPartnerHistory",
      "InputType": "provisioning.api.v1.GetPartnerHistoryRequest",
      "OutputType": "provisioning.api.v1.GetPartnerHistoryReply"
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "Service": "provisioning.api.v1.Partners",
      "Name": "GetPartnerResources",
      "InputType": "provisioning.api.v1.GetPartnerResourcesRequest",
      "OutputType": "provisioning.api.v1.GetPartnerResourcesReply"
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "Service": "provisioning.api.v1.Partners",
      "Name": "InitializeSFTPContainer",
      "InputType": "provisioning.api.v1.InitializeSFTPContainerRequest",
      "OutputType": "provisioning.api.v1.InitializeSFTPContainerReply"
    },
    {
      "Proto": "Protos/scheduling.api.v1.proto",
      "Service": "scheduling.api.v1.Greeter",
      "Name": "SayHello",
      "InputType": "scheduling.api.v1.HelloRequest",
      "OutputType": "scheduling.api.v1.HelloReply"
    },
    {
      "Proto": "Protos/subscription.api.v1.proto",
      "Service": "subscription.api.v1.Greeter",
      "Name": "SayHello",
      "InputType": "subscription.api.v1.HelloRequest",
      "OutputType": "subscription.api.v1.HelloReply"
    }
  ],
  "Fields": [
    {
      "Proto": "Protos/catalog.api.v1.proto",
      "MessageType": "catalog.api.v1.HelloRequest",
      "Name": "name",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/catalog.api.v1.proto",
      "MessageType": "catalog.api.v1.HelloReply",
      "Name": "message",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.CreatePartnerRequest",
      "Name": "partner",
      "Type": "identity.api.v1.PartnerModel",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.CreatePartnerReply",
      "Name": "id",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.DeletePartnerRequest",
      "Name": "id",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "id",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "name",
      "Type": "String",
      "FieldNumber": 2,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "contactName",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 3,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "contactPhone",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 4,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "contactEmail",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 5,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "publicKey",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 6,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerModel",
      "Name": "normalizedName",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 7,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerView",
      "Name": "id",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerView",
      "Name": "name",
      "Type": "String",
      "FieldNumber": 2,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerView",
      "Name": "contactName",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 3,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerView",
      "Name": "contactPhone",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 4,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.PartnerView",
      "Name": "contactEmail",
      "Type": "google.protobuf.StringValue",
      "FieldNumber": 5,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.GetPartnersReply",
      "Name": "partners",
      "Type": "identity.api.v1.PartnerView",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": true,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.GetPartnerRequest",
      "Name": "id",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/identity.partners.v1.proto",
      "MessageType": "identity.api.v1.GetPartnerReply",
      "Name": "partner",
      "Type": "identity.api.v1.PartnerView",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.ProvisionPartnerRequest",
      "Name": "partnerId",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.ProvisionPartnerRequest",
      "Name": "partnerName",
      "Type": "String",
      "FieldNumber": 2,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.ProvisionPartnerReply",
      "Name": "partnerId",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.RemovePartnerRequest",
      "Name": "partnerId",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.InitializeSFTPContainerReply",
      "Name": "succeeded",
      "Type": "Bool",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.RemovePartnerReply",
      "Name": "partnerId",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.GetPartnerHistoryRequest",
      "Name": "partnerId",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.GetPartnerHistoryReply",
      "Name": "partnerId",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.GetPartnerHistoryReply",
      "Name": "events",
      "Type": "provisioning.api.v1.ProvisioningEventView",
      "FieldNumber": 2,
      "IsRequired": false,
      "IsRepeated": true,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.ProvisioningEventView",
      "Name": "partnerId",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.ProvisioningEventView",
      "Name": "eventType",
      "Type": "String",
      "FieldNumber": 2,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.ProvisioningEventView",
      "Name": "sequence",
      "Type": "Int32",
      "FieldNumber": 3,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.ProvisioningEventView",
      "Name": "started",
      "Type": "String",
      "FieldNumber": 4,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.ProvisioningEventView",
      "Name": "completed",
      "Type": "String",
      "FieldNumber": 5,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.ProvisioningEventView",
      "Name": "succeeded",
      "Type": "Bool",
      "FieldNumber": 6,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.ProvisioningEventView",
      "Name": "errorMsg",
      "Type": "String",
      "FieldNumber": 7,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.GetPartnerResourcesRequest",
      "Name": "partnerId",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.GetPartnerResourcesReply",
      "Name": "partnerId",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.GetPartnerResourcesReply",
      "Name": "resources",
      "Type": "provisioning.api.v1.PartnerResourceView",
      "FieldNumber": 2,
      "IsRequired": false,
      "IsRepeated": true,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.PartnerResourceView",
      "Name": "resourceType",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.PartnerResourceView",
      "Name": "name",
      "Type": "String",
      "FieldNumber": 3,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.PartnerResourceView",
      "Name": "link",
      "Type": "Bool",
      "FieldNumber": 4,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.PartnerResourceView",
      "Name": "sensitive",
      "Type": "Bool",
      "FieldNumber": 5,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/provisioning.partners.v1.proto",
      "MessageType": "provisioning.api.v1.PartnerResourceView",
      "Name": "displayValue",
      "Type": "String",
      "FieldNumber": 6,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/scheduling.api.v1.proto",
      "MessageType": "scheduling.api.v1.HelloRequest",
      "Name": "name",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/scheduling.api.v1.proto",
      "MessageType": "scheduling.api.v1.HelloReply",
      "Name": "message",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/subscription.api.v1.proto",
      "MessageType": "subscription.api.v1.HelloRequest",
      "Name": "name",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    },
    {
      "Proto": "Protos/subscription.api.v1.proto",
      "MessageType": "subscription.api.v1.HelloReply",
      "Name": "message",
      "Type": "String",
      "FieldNumber": 1,
      "IsRequired": false,
      "IsRepeated": false,
      "IsMap": false,
      "ContainingOneof": null
    }
  ],
  "EnumValues": []
}